src/Analyzers/Core/Analyzers/RemoveUnnecessarySuppressions/AbstractRemoveUnnecessaryPragmaSuppressionsDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,558 @@
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Reflection;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeQuality;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+
+#if !NETCOREAPP
+using Roslyn.Utilities;
+#endif
+
+namespace Microsoft.CodeAnalysis.RemoveUnnecessarySuppressions
+{
+    internal abstract class AbstractRemoveUnnecessaryPragmaSuppressionsDiagnosticAnalyzer
+        : AbstractCodeQualityDiagnosticAnalyzer, IPragmaSuppressionsAnalyzer
+    {
+        private static readonly LocalizableResourceString s_localizableRemoveUnnecessarySuppression = new LocalizableResourceString(
+           nameof(AnalyzersResources.Remove_unnecessary_suppression), AnalyzersResources.ResourceManager, typeof(AnalyzersResources));
+        internal static readonly DiagnosticDescriptor s_removeUnnecessarySuppressionDescriptor = CreateDescriptor(
+            IDEDiagnosticIds.RemoveUnnecessarySuppressionDiagnosticId, s_localizableRemoveUnnecessarySuppression, s_localizableRemoveUnnecessarySuppression, isUnnecessary: true);
+
+        private readonly Lazy<ImmutableHashSet<int>> _lazySupportedCompilerErrorCodes;
+
+        protected AbstractRemoveUnnecessaryPragmaSuppressionsDiagnosticAnalyzer()
+            : base(ImmutableArray.Create(s_removeUnnecessarySuppressionDescriptor), GeneratedCodeAnalysisFlags.None)
+        {
+            _lazySupportedCompilerErrorCodes = new Lazy<ImmutableHashSet<int>>(() => GetSupportedCompilerErrorCodes());
+        }
+
+        protected abstract string CompilerErrorCodePrefix { get; }
+        protected abstract int CompilerErrorCodeDigitCount { get; }
+        protected abstract ISyntaxFacts SyntaxFacts { get; }
+        protected abstract (Assembly assembly, string typeName) GetCompilerDiagnosticAnalyzerInfo();
+
+        private ImmutableHashSet<int> GetSupportedCompilerErrorCodes()
+        {
+            try
+            {
+                // Use reflection to fetch compiler diagnostic IDs that are supported in IDE live analysis.
+                // Note that the unit test projects have IVT access to compiler layer, and hence can access this API.
+                // We have unit tests that guard this reflection based logic and will fail if the API is changed
+                // without updating the below code.
+
+                var (assembly, compilerAnalyzerTypeName) = GetCompilerDiagnosticAnalyzerInfo();
+                var compilerAnalyzerType = assembly.GetType(compilerAnalyzerTypeName)!;
+                var methodInfo = compilerAnalyzerType.GetMethod("GetSupportedErrorCodes", BindingFlags.Instance | BindingFlags.NonPublic)!;
```

Comment:
do we have a test that will catch things if this fails in teh future (i.e. if compiler change sthis)?  Also, I really don't like us using reflection into the compiler.  To the point that this is currently blocking for me.

Reply:
Yes, all unit tests validating unnecessary pragma suppressions for compiler diagnostics can be detected and removed with this analyzer/fixer will fail.